### PR TITLE
Display wheel version in the status page

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

Display wheel version in the status page
    
We now collect the wheel version from python checks and display it in the status page. This is mandatory to allow user to manually update a wheel and be able to see what is currently installed.

### CODE REVIEW

File: pkg/collector/check/check.go [hunk 1]
```diff
@@ -27,3 +27,9 @@ type Check interface {
 	GetWarnings() []error                                // return the last warning registered by the check
 	GetMetricStats() (map[string]int64, error)           // get metric stats from the sender
 }
+
+// PythonCheck extends the Check interface with specific method for the python checks
+type PythonCheck interface {
+	Check
+	Version() string // return the version of the check if available
+}
```
Comment (masci) [inline pkg/collector/check/check.go]: would it be more future proof to add `Version` to `Check` and just return the agent version for core checks? This way you'd also save a type assertion when creating the `Stats`.

File: pkg/status/helpers.go [hunk 2]
```diff
@@ -95,11 +95,7 @@ func FormatUnixTime(unixTime float64) string {
 }
 
 func printDashes(s string, dash string) string {
-	var dashes string
-	for i := 0; i < stringLength(s); i++ {
-		dashes += dash
-	}
-	return dashes
+	return strings.Repeat(dash, stringLength(s))
```
Comment (masci) [inline pkg/status/helpers.go]: 💥

File: pkg/status/dist/templates/collector.tmpl [hunk 3]
```diff
@@ -15,8 +15,8 @@ Collector
   {{end -}}
 
   {{- range .Checks}}
-    {{.CheckName}}
-    {{printDashes .CheckName "-"}}
+    {{.CheckName}}{{ if .CheckVersion }} ({{.CheckVersion}}){{ end }}
+    {{printDashes .CheckName "-"}}{{- if .CheckVersion }}{{printDashes .CheckVersion "-"}}---{{ end }}
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/status/dist/templates/collector.tmpl]: For the parenthesis and the space

File: releasenotes/notes/wheel-version-status-page-5d15e8241bbf20e1.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Display wheel version in the agent status page (for python checks).
```
Comment (masci) [inline releasenotes/notes/wheel-version-status-page-5d15e8241bbf20e1.yaml]: I'd rephrase to something like "Display the version for Python checks on the status page"

File: pkg/collector/check/check.go (hunk 1 above)
Comment (hush-hush) [reply pkg/collector/check/check.go]: Good question. On one side that force us to extend the API for every check in the agent and it would display the agent version for every Go check in the status page (is it useful ?). On the other side we avoid a type assertion and a new interface. 

If feel that having a `PythonCheck` interface could be useful going forward (a python check is more complex than a go one: wheel and stuff) and the type assertion is only done at the check creation, so no impact on performances.

I don't have a strong opinion, so if people think extending the `Check` interface is better I'll do it.

Comment (truthbk) [reply pkg/collector/check/check.go]: I kind of agree with the notion that `Version` being an attribute of all checks, not just python checks is more future-proof. It's also true the `PythonCheck` interface might make sense in the future, but maybe not specifically for this particular case. I don't have a die-for strong opinion, but I agree with @masci argument.

File: pkg/collector/py/loader.go [hunk 5]
```diff
@@ -97,6 +97,45 @@ func (cl *PythonCheckLoader) Load(config integration.Config) ([]check.Check, err
...
+				// display a meaningful error message.
+
+				typeName := "unable to detect type"
+				pyType := wheelVersionPy.Type()
+				if pyType != nil {
+					defer pyType.DecRef()
+					pyTypeStr := pyType.Str()
+					if pyTypeStr != nil {
+						defer pyTypeStr.DecRef()
+						typeName = python.PyString_AS_STRING(pyType.Str())
```
Comment (truthbk) [inline pkg/collector/py/loader.go]: I think we should do:
```
typeName = python.PyString_AS_STRING(pyTypeStr)
```

Otherwise we'll be creating a new reference after the `pyType.Str()` and never decreasing it. Beside, we already have the string object in `pyTypeStr`.

File: pkg/status/helpers.go (hunk 2 above)
Comment (truthbk) [reply pkg/status/helpers.go]: 🍰

**Review by truthbk — CHANGES_REQUESTED** [review]
I think I found a missing RefDec

File: pkg/collector/py/loader.go (hunk 5 above)
Comment (hush-hush) [reply pkg/collector/py/loader.go]: :man_facepalming: ...

**Review by truthbk — COMMENTED** [review]
This is good to go from a technical standpoint. I would like us to discuss what we should do about the versions of the long-running checks. I mean (other than `jmxfetch`), this doesn't really apply to `linux` or `windows`, only `macOS` and other less supported platforms - so it's not a huge deal and we can always fix later. But I think it's worth discussing (and addressing if necessary) now. What do you think?

File: pkg/collector/py/loader.go [hunk 6]
```diff
@@ -97,6 +97,45 @@ func (cl *PythonCheckLoader) Load(config integration.Config) ([]check.Check, err
 
 	// Try to find a class inheriting from AgentCheck within the module
 	checkClass, err := findSubclassOf(cl.agentCheckClass, checkModule, glock)
+
+	wheelVersion := ""
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/collector/py/loader.go]: I like `unversioned` since it will only apply to custom checks.

File: pkg/collector/corechecks/embed/apm.go [hunk 7]
```diff
@@ -43,6 +43,10 @@ func (c *APMCheck) String() string {
 	return "APM Agent"
 }
 
+func (c *APMCheck) Version() string {
+	return ""
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/collector/corechecks/embed/apm.go]: This is a good question ! I hesitated with executing trace/jmx/process to get their version.

That would mean executing the binary twice (once for the version and once to run it) and I don't know if it would be useful. Their version should be sync with the agent and are easily accessible by the user.

In the end I didn't do it because there is no way for the user to update them (since we don't provide packages for them). So I don't think it's worth the effort.

What do you think ?

[the remark this answers is not in the document]
Comment (truthbk) [reply pkg/collector/corechecks/embed/apm.go]: OK, so since the versions are in-sync with the agent, I think that makes sense. 

With `JMXFetch` I think it would help as it's definitely not aligned. I think I have an idea for JMXFetch using the release.json and `-X` ldflag that would probably work nicely. But I'd rather do it in a separate PR (I can work on that).

File: pkg/collector/corechecks/embed/jmx/check.go [hunk 8]
```diff
@@ -60,6 +60,10 @@ func (c *JMXCheck) String() string {
 	return c.name
 }
 
+func (c *JMXCheck) Version() string {
```
[the remark this answers is not in the document]
Comment (truthbk) [reply pkg/collector/corechecks/embed/jmx/check.go]: We'll address this separately 👍

**Review by truthbk — APPROVED** [review]
I approve this ~message~ PR! 😄
